internal/certgen/certgen.go

```diff
@@ -73,9 +86,40 @@ type keyPair struct {
 	Key        *certData
 }
 
-type certData struct {
-	Filename string
-	Data     []byte
+// writeTLSYAML writes out Kubernetes Secret YAML for a keypair
+// in a Secret of type `kubernetes.io/tls`. This prescribes the
+// filenames/keynames for the cert and key.
+func (kp *keyPair) writeTLSYAML(outputDir, namespace string) error {
+	secret := newTLSSecret(kp.SecretName, namespace, kp.Key.Data, kp.Cert.Data)
+	return writeSecret(outputDir+"/"+kp.SecretName+".yaml", secret)
+}
+
+// writeTLSKube writes a TLS Secret of the keypair out to Kubernetes.
+func (kp *keyPair) writeTLSKube(namespace string, client *kubernetes.Clientset) error {
+	secret := newTLSSecret(kp.SecretName, namespace, kp.Key.Data, kp.Cert.Data)
+	_, err := client.CoreV1().Secrets(namespace).Create(secret)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("secret/%s created\n", kp.SecretName)
+	return nil
+}
+
+// writePEMs writes both certificates of a keypair out to outputDir.
+func (kp *keyPair) writePEMs(outputDir string) error {
+	err := kp.Cert.writePEM(outputDir)
+	if err != nil {
+		return err
+	}
+	return kp.Key.writePEM(outputDir)
+}
+
+// ContourCerts holds all three keypairs required, the CA, the Contour, and Envoy keypairs.
+type ContourCerts struct {
+	CA               *keyPair
```

Comment:
I feel strongly that CA should not be a keyPair, there should be no reason that the private Key of the CA that signed the Envoy and Contour certs should not be forgotten as soon as those certs have been signed. This will ensure that nobody can use the CA in the contour namespace to mint additional certs.